### DISCUSSION TITLE
카테고리형, 연속형 feature 임베딩 방식 질문입니다!

### DISCUSSION THREAD

**Discussion by GY-Jeong** [bcaitech1/p4-dkt-lastffang#41] [Q&A]
## 다른점 1

지금 master branch 코드의 model.py 에서는

https://github.com/bcaitech1/p4-dkt-lastffang/blob/6ab84ed0c4ebb49541389523e47974ad55be1d9e/dkt/model.py#L28-L34

에서 처럼 연속형 feature (정답코드, 0 1 2)도 카테고리형 feature처럼 embedding을 거치는 것을 볼 수 있습니다!

저희가 강의시간에 다룬 연속형 feature를 다루는 방법은

![image](https://user-images.githubusercontent.com/28976984/120021472-134c7480-c026-11eb-8e06-4c339d277aa9.png)

이런 방식처럼 그대로 linear layer에 넣어버리는 방식이었는데, 저렇게 연속형 feature도 embedding한다면, 각 feature마다 max값을 원하는 dimension으로 바꿔줘야 합니다.

`nn.Embedding(max값+1, self.hidden_dim//3)` 이렇게요!

대분류 (1~9)의 값도 https://github.com/bcaitech1/p4-dkt-lastffang/tree/large_category 에서는 embedding으로

https://github.com/bcaitech1/p4-dkt-lastffang/blob/792d0e6e1d05650632ba1824b3ab2b351d783485/dkt/model.py#L25

이런 식으로 넘겨줬는데 어떤 방식을 사용하는 것이 좋을까요?

## 다른점 2

지금 코드에서는 모든 feature가 연속형으로 간주되어 동일한 size로 변환 뒤 concat하게 됩니다.

강의에서 다룬 내용중 하나는, 연속형과 카테고리형을 hidden size의 반씩으로 projection시킨 뒤, concat하는 방식이었는데 저희 코드를 수정할 필요가 있을까요? 아니면 지금처럼 모든 feature를 동일한 size로 변환하고 concat하는 것이 좋을까요?

결국 실험을 해서 결과를 보고 판단해야 할 것 같기는 한데, 글만 보고 생각하시는 의견이 있으신지 궁금합니다!

**Comment by gonipark** [comment]
다른점1
위에서 말씀하신 (정답코드, 0 1 2)는 연속형 변수가 아니라 카테고리형 feature로 취급(?)하고 있기 때문에 임베딩을 거친다고 생각합니다. 그래서 large_category 은 근영님이 사용하신 방식이 맞다고 생각해요!
다른점2
이건 말씀하신거처럼 정말 예측 불가하다고 생각합니다. 데이터가 그렇게 많지 않아서, dimension을 줄일 필요성이 크지 않으니 dimension을 그대로 가지고 가도 되지 않을까 생각하고 있긴 해요. 그 차이를 수치화하는걸 담주 목표로 해도 좋을 것 같아요!
